### PR TITLE
WIP: handle task logic in ChallengeDetailActivity // clean-architecture

### PR DESCRIPTION
**Pull request by negue**
In order to be able to use the task button (habit +/-, todo etc) in the ChallengeDetailActivity, we need to refactor how we fire up those api calls, save the new data or even play the sounds.

Most of the boilerplate is from the clean-architecture branch.

Current changes:
- private ApiService
- renamed ApiHelper to IApiClient
- ApiClient now handles `.compose(configureApiCallObserver())` calls, just returns the result without `HabitResponse`
- extracted multiple UseCases:
   - BuyReward
   - CheckClassSelection
   - ChecklistCheck
   - DailyCheck
   - DisplayItemDrop
   - HabitScore
   - LevelUp
   - NotifyUser
   - TodoCheck

### CODE REVIEW

**Comment by TheHollidayInn** [comment]
This looks great!! I'm super excited.

My only concern is for the directory structure of the data. 
- Should we do: data/repositories 
 - Or even: data/tasks
       - /local
       - /remote

File: Habitica/src/main/java/com/magicmicky/habitrpgwrapper/lib/api/ApiService.java [hunk 1]
```diff
@@ -100,6 +100,9 @@
     @GET("tasks/user")
     Observable<HabitResponse<TaskList>> getTasks();
 
+    @GET("tasks/user")
+    Observable<HabitResponse<ArrayList<Task>>> getUserTasks();
```
Comment (vIiRuS) [inline Habitica/src/main/java/com/magicmicky/habitrpgwrapper/lib/api/ApiService.java]: Why did you add this call a second time, with a different return type?

File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/UseCase.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * Builds an {@link rx.Observable} which will be used when executing the current {@link UseCase}.
+     */
+    protected abstract Observable<T> buildUseCaseObservable(Q requestValues);
+
+    @SuppressWarnings("unchecked")
+    public Observable<T> observable(Q requestValues) {
+        return this.buildUseCaseObservable(requestValues)
+                .subscribeOn(postExecutionThread.getScheduler())
```
Comment (vIiRuS) [inline Habitica/src/main/java/com/habitrpg/android/habitica/interactors/UseCase.java]: this causes all usecases to be run on the main thread (since that's the postexecutionthread we're using). In the clean-architecture branch, I did this because realm objects always have to be handled on the same thread and this was a working temporary solution. but we should probably find a way to at least have some of the usecases run on a background thread. Suggestions for how to implement this best are appreciated.

**Comment by vIiRuS** [comment]
@TheHollidayInn depends on if we want to put other things than the repositories in `data`. if not, I'd probably rename `data` to `repositories`. if we want to add other things, then going with `data/repositories` sounds good. I wouldn't split it further at this point, because I don't think we'll have more than a handfull of different repositories.

**Comment by TheHollidayInn** [comment]
Well, from the code above it looks like for each model we will have a DBFLow, local repositroy, and task repo implementation

**Comment by vIiRuS** [comment]
I wouldn't create a repository per model, but rather group them into sections. so that there might be one "Social" repository and a "Inventory" one. that each handle working with the matching data.
for each one we'll have something like this.

- Repository interface - data/Repository.java
- Repository implementation - data/impl/RepositoryImpl.java
- Local Repository interface - data/local/LocalRepository.java
- Local Repository implementation - data/local/impl/LocalRepositoryImpl.java

so while there are 4 files per repository, there is only 1 file in each folder.

**Comment by TheHollidayInn** [comment]
Hmm, I don't think I would do that. I think Challenges should have their own, Tasks, Users, Groups, etc.

**Comment by vIiRuS** [comment]
Not saying they shouldn't. but we p.e. don't need an "EggsRepository". that could go in an Inventory Repository or something like that. That's what I meant.

**Comment by TheHollidayInn** [comment]
Ah, yes. I agree with that.

File: Habitica/src/main/java/com/habitrpg/android/habitica/data/implementation/TaskRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.ArrayList;
+
+import rx.Observable;
+
+
+public class TaskRepositoryImpl extends BaseRepositoryImpl<TaskLocalRepository> implements TaskRepository {
+
+
+    private APIHelper apiHelper;
```
Comment (negue) [inline Habitica/src/main/java/com/habitrpg/android/habitica/data/implementation/TaskRepositoryImpl.java]: yeah I guess it would be better to use only the `apiclient` , but I'd like to implement the same interface of `ApiService` which would result that `configureApiCallObserver`  has to return `HabitResponse<T>` again

or we'd have two interfaces one with `HabitResponse<T>` and one without, what do you think? @vIiRuS

File: Habitica/src/main/java/com/magicmicky/habitrpgwrapper/lib/api/ApiService.java (hunk 1 above)
Comment (negue) [inline Habitica/src/main/java/com/magicmicky/habitrpgwrapper/lib/api/ApiService.java]: I guess I thought at first thats easier to use the ArrayList<Task>than a TaskList (which has all Tasks in a HashMap) , gonna revert this again^^

File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/UseCase.java (hunk 2 above)
Comment (negue) [inline Habitica/src/main/java/com/habitrpg/android/habitica/interactors/UseCase.java]: I'm gonna try to get an overview of a simple UseCase and what subscribe/observeOn are called and then decide what to add and what not.

My first suggestion would be at first: let the usecase handle if it wants to use the subscribeOn or not?

**Comment by negue** [comment]
@vIiRuS I added some more use-cases, should something like `EventBus.getDefault().post(new TaskUpdatedEvent(event.Task)); ` be added inside an use-case or stay inside a `.subscribe()`-call?

File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/CheckClassSelectionUseCase.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        bundle.putInt("hairBase", user.getPreferences().getHair().getBase());
+        bundle.putString("hairColor", user.getPreferences().getHair().getColor());
+        bundle.putInt("hairMustache", user.getPreferences().getHair().getMustache());
+        bundle.putInt("hairBeard", user.getPreferences().getHair().getBeard());
+        bundle.putBoolean("isInitialSelection", event.isInitialSelection);
+        bundle.putString("currentClass", event.currentClass);
+
+        Intent intent = new Intent(HabiticaApplication.currentActivity, ClassSelectionActivity.class);
+        intent.putExtras(bundle);
+        HabiticaApplication.currentActivity.startActivityForResult(intent, SELECT_CLASS_RESULT);
```
Comment (negue) [inline Habitica/src/main/java/com/habitrpg/android/habitica/interactors/CheckClassSelectionUseCase.java]: yeah its a bit too much, but the ClassSelection is used in multiple ways so it kinda "screams" to be a usecase 😄

File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/ListTasksUseCase.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Inject
+    public ListTasksUseCase(TaskRepository taskRepository, ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
+        super(threadExecutor, postExecutionThread);
+        this.taskRepository = taskRepository;
+    }
+
+    @Override
+    protected Observable<ArrayList<Task>> buildUseCaseObservable(ListTasksUseCase.RequestValues requestValues) {
+        return taskRepository.getTasks(requestValues.taskType);
```
Comment (negue) [inline Habitica/src/main/java/com/habitrpg/android/habitica/interactors/ListTasksUseCase.java]: `ListTasksUseCase` was a test UseCase-Class, gonna remove it
